File: CheckImageView/app/src/main/java/deadline/checkImageView/CheckImageView.java
package deadline.checkImageView;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.AttributeSet;
import android.widget.ImageView;

/**
 * @author deadline
 */
public class CheckImageView extends ImageView implements ICheckAttacher{

    private CheckAttacher mAttacher;

    public CheckImageView(Context context) {
        this(context, null);
    }

    public CheckImageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CheckImageView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setup();
    }

    private void setup() {
        setScaleType(ScaleType.MATRIX);
        mAttacher = new CheckAttacher(this);
    }

    @Override
    public boolean isChecked() {
        return mAttacher.isChecked();
    }

    @Override
    public void setChecked(boolean checked) {
        mAttacher.setChecked(checked);
    }

    @Override
    public void setCheckedScale(float scale) {
        mAttacher.setCheckedScale(scale);
    }

    @Override
    public float getScale() {
        return mAttacher.getScale();
    }

    @Override
    public void setZoomAnimationDuration(long duration) {
        mAttacher.setZoomAnimationDuration(duration);
    }

    @Override
    public void setOnCheckChangeListener(CheckAttacher.OnCheckedChangeListener listener) {
        mAttacher.setOnCheckChangeListener(listener);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mAttacher.moveToCenter();
    }

    @Override
    // setImageBitmap calls through to this method
    public void setImageDrawable(Drawable drawable) {
        super.setImageDrawable(drawable);
        if (null != mAttacher) {
            mAttacher.moveToCenter();
        }
    }

    @Override
    public void setImageResource(int resId) {
        super.setImageResource(resId);
        if (null != mAttacher) {
            mAttacher.moveToCenter();
        }
    }

    @Override
    public void setImageURI(Uri uri) {
        super.setImageURI(uri);
        if (null != mAttacher) {
            mAttacher.moveToCenter();
        }
    }

    @Override
    protected boolean setFrame(int l, int t, int r, int b) {
        boolean changed = super.setFrame(l, t, r, b);
        if (null != mAttacher) {
            mAttacher.moveToCenter();
        }
        return changed;
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mAttacher.release();
    }
}

File: CheckImageView/app/src/main/java/kot/checkImageView/CheckImageView.kt
package kot.checkImageView

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.support.annotation.Nullable
import android.util.AttributeSet
import android.widget.ImageView

/**
 * @author deadline
 * @time 2017/10/14
 */
class CheckImageView : ImageView, ICheckAttacher {

    private var mAttacher: CheckAttacher = CheckAttacher(this)

    constructor(context: Context?) : super(context, null)

    constructor(context: Context?, @Nullable attrs: AttributeSet) : super(context, attrs, 0)

    /* command + option + / 注释 */
    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
    }

    override fun isChecked(): Boolean {
        return mAttacher.isChecked()
    }

    override fun setChecked(checked: Boolean) {
        mAttacher.setChecked(checked)
    }

    override fun setCheckedScale(scale: Float) {
        mAttacher.setCheckedScale(scale)
    }

    override fun getScale(): Float {
        return mAttacher.getScale()
    }

    override fun setZoomAnimationDuration(duration: Long) {
        mAttacher.setZoomAnimationDuration(duration)
    }

    override fun setOnCheckChangeListener(listener: OnCheckedChangeListener) {
        mAttacher.setOnCheckChangeListener(listener)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        mAttacher.moveToCenter()
    }

    override // setImageBitmap calls through to this method
    fun setImageDrawable(drawable: Drawable?) {
        super.setImageDrawable(drawable)
        mAttacher.moveToCenter()
    }

    override fun setImageResource(resId: Int) {
        super.setImageResource(resId)
        mAttacher.moveToCenter()
    }

    override fun setImageURI(uri: Uri?) {
        super.setImageURI(uri)
        mAttacher.moveToCenter()
    }

    override fun setFrame(l: Int, t: Int, r: Int, b: Int): Boolean {
        val changed = super.setFrame(l, t, r, b)
        mAttacher.moveToCenter()
        return changed
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mAttacher.release()
    }
}